lib/smack/SmackRep.cpp

```diff
@@ -181,19 +222,88 @@ const Stmt* SmackRep::memset(const llvm::MemSetInst& msi) {
   stringstream name;
   vector<const Expr*> args;
   name << "$memset." << region;
-  for (unsigned i = 0; i < msi.getNumOperands() - 1; i++)
+  args.push_back(expr(msi.getOperand(0)));
+  args.push_back(expr(msi.getOperand(1)));
+  const llvm::Value* setSize = msi.getOperand(2);
+  args.push_back((getIntSize(setSize) == 32)? Expr::fn("$zext.i32.ref", expr(setSize)): expr(setSize));
+  for (unsigned i = 3; i < msi.getNumOperands() - 1; i++)
     args.push_back(expr(msi.getOperand(i)));
   return Stmt::call(name.str(),args);
 }
 
-const Expr* SmackRep::pa(const Expr* base, int index, int size) {
-  return pa(base, Expr::lit(index), Expr::lit(size));
+const Stmt* SmackRep::load(const llvm::Value* addr, const llvm::Value* val) {
+// The tricky part is that we could expr(li) is actually expr(val) so that it is possible to pass li to val. 
+  const Expr* rhs = mem(addr);
+
+  if (isFloat(val))
+    rhs = Expr::fn("$si2fp", rhs);
+
+  if (!SmackOptions::BitVectors || (SmackOptions::InferFieldOverlap && isFieldDisjoint(addr, llvm::cast<const llvm::Instruction>(val))))
```

Comment:
The conversion `$si2fp` is not applied in the second branch. This should be rewritten as follows:
````
Expr* rhs;
if (!SmackOptions::BitVectors || …) {
  rhs = mem(addr);
} else {
  rhs = …;
}
if (isFloat(val))
  rhs = Expr::fn("$si2fp", rhs);
return Stmt::assign(expr(val), rhs);
````
In fact the `$si2fp` function should probably depend on `addr`'s element type and the type of `val`.